addon/components/horizontal-bar-chart.js

```diff
@@ -451,12 +455,41 @@ const HorizontalBarChartComponent = ChartComponent.extend(FloatingTooltipMixin,
         return this._getElementForValue(valueLabelElements, val);
       });
 
-      const leftLabels = negativeValueLabels.concat(positiveGroupingLabels);
-      const rightLabels = positiveValueLabels.concat(negativeGroupingLabels);
+      const maxNegativeValueLabelWidth = this._maxWidthOfElements(negativeValueLabels);
+      const maxPositiveGroupingLabelWidth = Math.min(maxLabelWidth, this._maxWidthOfElements(positiveGroupingLabels));
+
+      const maxPositiveValueLabelWidth = this._maxWidthOfElements(positiveValueLabels);
+      const maxNegativeGroupingLabelWidth = Math.min(maxLabelWidth, this._maxWidthOfElements(negativeGroupingLabels));
+
+      const minValue = Math.abs(this.get('minValue'));
+      const maxValue = this.get('maxValue');
+      const containerWidth = this.get('outerWidth');
+      const axisTitleOffset = this.get('yAxisTitleHeightOffset') + 5;
+      const chartPadding = 2 * this.get('labelPadding') + axisTitleOffset;
+      const chartWidth = containerWidth - maxNegativeValueLabelWidth - maxPositiveValueLabelWidth - chartPadding;
+
+      const xScale = this.xScaleForWidth(chartWidth);
+      const maxNegativeBarWidth = Math.abs(xScale(minValue) - xScale(0));
+      const maxPositiveBarWidth = Math.abs(xScale(maxValue) - xScale(0));
+
+      const leftGroupLabelWidth = maxPositiveGroupingLabelWidth;
+      const leftChartWidth = maxNegativeValueLabelWidth + maxNegativeBarWidth;
+
+      const rightGroupLabelWidth = maxNegativeGroupingLabelWidth;
+      const rightChartWidth = maxPositiveValueLabelWidth + maxPositiveBarWidth;
+
+      var barLength;
+      var leftWidth = maxNegativeValueLabelWidth;
+      var rightWidth = maxPositiveValueLabelWidth;
+
+      if (leftGroupLabelWidth > leftChartWidth) {
+        barLength = (containerWidth - leftGroupLabelWidth - rightWidth - chartPadding) * minValue / maxValue;
+        leftWidth = Math.min(leftGroupLabelWidth - barLength, containerWidth - rightWidth - chartPadding);
+      } else if (rightGroupLabelWidth > rightChartWidth) {
+        barLength = (containerWidth - rightGroupLabelWidth - leftWidth - chartPadding) * maxValue / minValue;
+        rightWidth = Math.min(rightGroupLabelWidth - barLength, containerWidth - leftWidth - chartPadding);
+      }
```

Comment:
This logic is incredibly difficult to follow. Additionally, this whole `else` statement is pretty long. I think the intent would be clearer if this were broken up into named helper functions.